sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1909,458 @@ def transform(self, X):
+        """Build the transform functions."""
+        check_is_fitted(self, 'quantiles_')
+
+        if self.ignore_implicit_zeros:
+            warnings.warn("'ignore_implicit_zeros' takes effect only with"
+                          " sparse matrix. This parameter has no effect.")
+
+        references = np.linspace(0, 1, self.n_quantiles, endpoint=True)
+
+        self._f_transform = tuple([
+            interp1d(quantiles_feature, references,
+                     bounds_error=False,
+                     fill_value=0.)
+            for quantiles_feature in self.quantiles_.T])
+
+        self._f_inverse_transform = tuple([
+            interp1d(references, quantiles_feature,
+                     bounds_error=False,
+                     fill_value=0.)
+            for quantiles_feature in self.quantiles_.T])
+
+    def _dense_fit(self, X):
+        """Compute percentiles for dense matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The data used to scale along the features axis.
+        """
+        rng = check_random_state(self.random_state)
+
+        # subsample the matrix X if necessary
+        n_samples, n_features = X.shape
+        if self.subsample < n_samples:
+            subsample_idx = rng.permutation(range(n_samples))[:self.subsample]
+        else:
+            subsample_idx = range(n_samples)
+
+        # for compatibility issue with numpy<=1.8.X, references
+        # need to be a list scaled between 0 and 100
+        references = np.linspace(0, 1, self.n_quantiles,
+                                 endpoint=True).tolist()
+        # references is a list that we need to scale between
+        # 0 and 100.
+        self.quantiles_ = np.array([np.percentile(
+            X[subsample_idx, feature_idx], [x * 100 for x in references])
+                                    for feature_idx in range(n_features)]).T
+
+    def _sparse_fit(self, X):
+        """Compute percentiles for sparse matrices.
+
+        Parameters
+        ----------
+        X : sparse matrix CSC, shape (n_samples, n_features)
+            The data used to scale along the features axis. The sparse matrix
+            needs to be semi-positive.
+        """
+        rng = check_random_state(self.random_state)
+
+        n_samples, n_features = X.get_shape()
+
+        # for compatibility issue with numpy<=1.8.X, references
+        # need to be a list
+        references = np.linspace(0, 1, self.n_quantiles,
+                                 endpoint=True).tolist()
+        # references is a list that we need to scale between
+        # 0 and 100.
+        self.quantiles_ = []
+        for feature_idx in range(n_features):
+            column_nnz_data = X.data[X.indptr[feature_idx]:
+                                     X.indptr[feature_idx + 1]]
+            if len(column_nnz_data) > self.subsample:
+                column_subsample = (self.subsample * len(column_nnz_data) //
```

Comment:
For very sparse features, this means very few samples (although then this might be a bad scaling strategy).

Can we analytically account for the zeros when `ignore_implicit_zeros == False`, instead of needing to materialise an array containing the zeros?